core/src/main/java/net/opentsdb/query/BaseQueryNodeConfig.java

```diff
@@ -183,10 +187,60 @@ public boolean hasKey(final String key) {
   }
   
   @Override
-  public abstract boolean equals(final Object o);
-  
+  public boolean equals(final Object o) {
+    if (this == o)
+      return true;
+    if (o == null || getClass() != o.getClass())
+      return false;
+
+    final BaseQueryNodeConfig queryConfig = (BaseQueryNodeConfig) o;
+
+    final boolean result = Objects.equal(getType(), queryConfig.getType())
+            && Objects.equal(id, queryConfig.getId());
+    if (!result) {
+      return false;
+    }
+
+
+    if (!Comparators.ListComparison.equalLists(sources, queryConfig.getSources())) {
+      return false;
+    }
+
+    return true;
+
+  }
+
+
+
   @Override
-  public abstract int hashCode();
+  public int hashCode() {
+    return buildHashCode().asInt();
+  }
+
+
+  /** @return A HashCode object for deterministic, non-secure hashing */
+  public HashCode buildHashCode() {
+    final HashCode hc = Const.HASH_FUNCTION().newHasher()
+            .putString(Strings.nullToEmpty(id), Const.UTF8_CHARSET)
+            .putString(Strings.nullToEmpty(type), Const.UTF8_CHARSET)
+            .hash();
+    final List<HashCode> hashes =
+            Lists.newArrayListWithCapacity(2);
+
+    hashes.add(hc);
+
+    if (sources != null) {
```

Comment:
For these cases we want to avoid creating a second hasher object and higher level `hashes` array to save memory. Instead of `hc` make it the hasher and don't call `.hash()` yet. Instead in the loop after sorting, just call the `hasher.putString(...)` method on the original hash. The `hashes` list pattern is only needed if we have a complex sub object that returns a `HashCode`.